### PR TITLE
Add ability for `verdi process report` to automatically query last node

### PR DESCRIPTION
**Pull request by ljbeal**
Adds the `-e` and `--end` flags to `verdi process report`, replacing the requirement of specifiying a node.

Instead, report will query `verdi process list -a` and call on the last `pk` found. This makes the process reproducible in a single line, and removes a step in debugging a calculation.

This feature also attempts to collect the remote dir of the node, for further debugging.

It's highly probable that there's a cleaner way of implementing this, but it has been helpful for my workflows, feel free to advise (or go ahead any make) any changes to the implementation!

### CODE REVIEW

File: aiida/cmdline/commands/cmd_process.py [hunk 1]
```diff
@@ -185,11 +185,33 @@ def process_call_root(processes):
 @click.option(
     '-m', '--max-depth', 'max_depth', type=int, default=None, help='Limit the number of levels to be printed.'
 )
+@click.option(
+    '-e', '--end', 'last', is_flag=True, default=False, help='Report for the most recent node in verdi process list -a'
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_process.py:189]: Don't think that `--end` is really intuitive. Maybe something like the following would be more intuitive
```suggestion
    '-M', '--most-recent-node', 'most_recent_node', is_flag=True, default=False, help='Select the most recently created process node.'
```

I chose `-M` because `-m` is already in use. I think it would make sense to also add this option to `verdi process show` and `verdi process status`. These also don't use `-M` yet

File: aiida/cmdline/commands/cmd_process.py [hunk 2]
```diff
@@ -185,11 +185,33 @@ def process_call_root(processes):
...
+
+    if last:
+        # query all nodes, getting the last one available
+        builder = CalculationQueryBuilder()
+        lastnode_info = builder.get_projected(builder.get_query_set(), CalculationQueryBuilder.default_projections)[-1]
+        # take pk from node info and load that node
+        pk = lastnode_info[0]
+        node = load_node(pk)
+        # set process list to the singular final node
+        processes = [node]
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_process.py:206]: I wouldn't use the `CalculationQueryBuilder` here since that is more complicated than we need. We can use a simple `QueryBuilder` directly:
```suggestion
    if most_recent_node:
        builder =  QueryBuilder().append(ProcessNode, tag='n').order_by({'n': {'ctime': 'desc'}})
        processes = builder.first(flat=True)
```

File: aiida/cmdline/commands/cmd_process.py [hunk 3]
```diff
@@ -185,11 +185,33 @@ def process_call_root(processes):
...
+
+        echo.echo(f'Reporting for final node (pk: {pk})')
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_process.py:208]: Would make this an info because you might not always want to see this output. This way it does not show by default, but can be made visible with `-v INFO`
```suggestion
        echo.echo_info(f'Showing results for most recent node {node}')
```

File: aiida/cmdline/commands/cmd_process.py [hunk 4]
```diff
@@ -185,11 +185,33 @@ def process_call_root(processes):
...
+
+        try:
+            remote_dir = node.outputs.remote_folder.get_attribute('remote_path')
+            echo.echo(f'Remote Directory: {remote_dir}')
+        except AttributeError:
+            echo.echo(f'No Remote Directory Found')
```
Comment (sphuber) [inline aiida/cmdline/commands/cmd_process.py:214]: Why did you add this?

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @ljbeal . I think this could indeed be useful. I have added some suggestions for the implementation. I think it would also make sense to add this to `verdi process show` and `verdi process status`. If you agree, we can move the option definition to `aiida.cmdline.params.options.main` as a reusable option. We would also have to add some tests. I can give pointers if you'd like, or take over if you don't have the time

File: aiida/cmdline/commands/cmd_process.py (hunk 1 above)
Comment (ljbeal) [reply aiida/cmdline/commands/cmd_process.py:189]: I agree, I originally wanted `--last`, but the shorthand for that would be `-l`, which is ambiguous (l vs I),  and I associate this with `list`. Your suggestion is definitely better, though perhaps `--recent`/`-R` would be more intuitive? Or even `--last` and `-L`?

File: aiida/cmdline/commands/cmd_process.py (hunk 2 above)
Comment (ljbeal) [reply aiida/cmdline/commands/cmd_process.py:206]: Perfect, this seems to be a much more efficient implementation. I am still quite unfamiliar with the `QueryBuilder` family (clearly!)

File: aiida/cmdline/commands/cmd_process.py (hunk 4 above)
Comment (ljbeal) [reply aiida/cmdline/commands/cmd_process.py:214]: It's purely for debugging reasons, before I learned about `verdi calcjob gotocomputer`. This can either be removed or set to `echo_info`.

I still use it occasionally since it lets me `cat {remote_dir}/_aiidasubmit.sh` without leaving the current working dir, for example. But just like the `Reporting for final node` above, it's not strictly _required_ info.

**Comment by ljbeal** [comment]
Hi @sphuber, thanks for the response, and apologies for the delay in my own. The suggestions you made are good ones, I have responded in place.

I like the idea of making it a common variable, I found myself wanting to implement this for `verdi node show` at the very least. 

Indeed I was lazy with the tests, I'll get on that as a priority once I get `aiida-bigdft` replaced on PyPi. 

As for the pointers vs you doing the work, it's entirely up to you. I'm happy to do it myself, though if you choose to go ahead and make the necessary changes I'll definitely go take a look at the git diff. I get the sense that it will be useful info on the plugin side of things in either case.

File: aiida/cmdline/commands/cmd_process.py (hunk 1 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_process.py:189]: I find `--most-recent-node` way more intuitive than `--recent` because the latter doesn't imply a single node, but could be a few nodes, e.g. all nodes from last 24 hours. I can see how one could object to the longer option name, but that is why there is the shorthand _and_ we have tab completion of options.

File: aiida/cmdline/commands/cmd_process.py (hunk 2 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_process.py:206]: The `CalculationQueryBuilder` is simply a utility class that builds on top of the `QueryBuilder` functionality. It is mostly designed and used for the `verdi process list` command. If you need a simple query, your best bet is to go straight to the `QueryBuilder` yourself

File: aiida/cmdline/commands/cmd_process.py (hunk 4 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_process.py:214]: > It's purely for debugging reasons, before I learned about `verdi calcjob gotocomputer`. This can either be removed or set to `echo_info`.

I don't like it so much because it is specific to `CalcJob`s and doesn't apply to any other processes like calcfunctions, `WorkChains` etc.

> I still use it occasionally since it lets me `cat {remote_dir}/_aiidasubmit.sh` without leaving the current working dir, for example. 

This will only work if your `CalcJob` ran on the `localhost` though. If you ran it on a remote computer, this won't work and you would anyway have to use `verdi calcjob gotocomputer`.

> But just like the `Reporting for final node` above, it's not strictly _required_ info.

The difference is that this applies to all processes. In addition, it can actually help to prevent confusion. Since when you use this flag you think you might get the last node, but another one could have run in the meantime (if the daemon is running and has other proceses running), and so really you got another node than you were anticipating. In this case an explicit reference of exactly which node was matched can prevent confusion.

**Comment by sphuber** [comment]
> As for the pointers vs you doing the work, it's entirely up to you. I'm happy to do it myself, though if you choose to go ahead and make the necessary changes I'll definitely go take a look at the git diff. I get the sense that it will be useful info on the plugin side of things in either case.

For me there is no rush and I am happy to wait. It is just that I think this is a useful feature and so think it would be good to get it in, but if requests for adding tests block that because you don't have the time for it, I will happily pitch in.

**Comment by sphuber** [comment]
Superseded by #6283
